File: go/general/1025. 除数博弈.go
package main

func divisorGame(N int) bool {
	if N == 1 {
		return false
	}
	dp := make([]bool, N+1)
	dp[1] = false
	dp[2] = true

	for i := 3; i < N+1; i++ {
		dp[i] = false
		for x := 1; x < i; x++ {
			if i%x == 0 && dp[i-x] == false {
				dp[i] = true
				break
			}
		}
	}
	return dp[N]
}

File: python/动态规划/1025. 除数博弈.py
class Solution:
    def divisorGame(self, N: int) -> bool:
        if N == 1:
            return False

        # dp存储操作数为i时的玩家的获胜情况
        dp = [False] * (N+1)
        dp[1] = False
        dp[2] = True

        for i in range(3,N+1):
            for x in range(1,i):
                # 玩家都以最佳状态，即玩家操作i后的操作数i - x应尽可能使对手输，即dp[i - x]应尽可能为false
                # 所以遍历x = 1~i-1, 寻找x的约数，使得dp[i - x] = false
                # 那么dp[i] = true即当前操作数为i的玩家能获胜
                # 如果找不到则为false，会输掉

                if i%x == 0 and not dp[i-x]:
                    dp[i] =True
                    break
        return dp[N]